README.md
# gltf-viewer [![status](https://img.shields.io/badge/glTF-2%2E0-green.svg?style=flat)](https://github.com/KhronosGroup/glTF)
[![crates.io](https://img.shields.io/crates/v/gltf-viewer.svg)](https://crates.io/crates/gltf-viewer)
[![GitHub release](https://img.shields.io/github/release/bwasty/gltf-viewer.svg)](https://github.com/bwasty/gltf-viewer/releases/latest)
 [![](https://tokei.rs/b1/github/bwasty/gltf-viewer)](https://github.com/Aaronepower/tokei)
 [![Build Status](https://travis-ci.org/bwasty/gltf-viewer.svg?branch=master)](https://travis-ci.org/bwasty/gltf-viewer)
 [![Build status](https://ci.appveyor.com/api/projects/status/51ukh02thpb0r9cf/branch/master?svg=true)](https://ci.appveyor.com/project/bwasty/gltf-viewer/branch/master)
 [![Docker build status](https://img.shields.io/docker/build/bwasty/gltf-viewer.svg)](https://hub.docker.com/r/bwasty/gltf-viewer/tags/)
 [![Maintenance](https://img.shields.io/badge/maintenance-passively--maintained-yellowgreen.svg)](https://github.com/rust-lang/rfcs/blob/master/text/1824-crates.io-default-ranking.md#maintenance)

Rust [glTF 2.0](https://github.com/KhronosGroup/glTF) viewer, written using the [gltf](https://github.com/gltf-rs/gltf) crate and plain OpenGL.

**Current state**: All [official sample models](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0) can be loaded and are rendered with the [reference PBR shader](https://github.com/KhronosGroup/glTF-WebGL-PBR). Example: <br>
<img width="412" alt="SciFiHelmet" title="SciFiHelmet" src="https://user-images.githubusercontent.com/1647415/30771307-d70dbd26-a044-11e7-9ed1-b0e2ba80198c.png"><br>
Gallery with all sample models: https://bwasty.github.io/gltf-viewer/0.3.0/

Some glTF features are not yet implemented, most notably **animations**. See [#3](https://github.com/bwasty/gltf-viewer/issues/3) for details.

## Installation
### Binaries (Win/Linux/macOS)
See [Latest Release](https://github.com/bwasty/gltf-viewer/releases/latest)
### From crate (requires [Rust](https://www.rust-lang.org))
```shell
cargo install gltf-viewer
```
Latest version (unstable):
```shell
cargo install --git https://github.com/bwasty/gltf-viewer.git
```
## Usage
```
USAGE:
    gltf-viewer [OPTIONS] <FILE>

OPTIONS:
    -v, --verbose                    Enable verbose logging (log level INFO). Can be repeated up to 3 times to increase
                                     log level to DEBUG/TRACE)
    -s, --screenshot <FILE>          Create screenshot (PNG)
    -w, --width <WIDTH>              Width in pixels [default: 800]
    -h, --height <HEIGHT>            Height in pixels [default: 600]
    -c, --count <COUNT>              Saves N screenshots of size WxH, rotating evenly spaced around the object [default:
                                     1]
        --headless                   Use real headless rendering for screenshots (default is a hidden window)
                                     [EXPERIMENTAL - see README for details]
        --straight                   Position camera in front of model if using default camera (i.e. glTF doesn't
                                     contain a camera or `--cam-index -1` is passed).
        --scene <scene>              Index of the scene to load [default: 0]
        --cam-index <CAM-INDEX>      Use the glTF camera with the given index (starting at 0).
                                     Fallback if there is none: determine 'nice' camera position based on the scene's
                                     bounding box. Can be forced by passing -1.
                                     Note: All other camera options are ignored if this one is given. [default: 0]
        --cam-pos <CAM-POS>          Camera (aka eye) position override as comma-separated Vector3. Example: 1.2,3.4,5.6
        --cam-target <CAM-TARGET>    Camera target (aka center) override as comma-separated Vector3. Example:
                                     1.2,3.4,5.6
        --cam-fovy <CAM-FOVY>        Vertical field of view ('zoom') in degrees. [default: 75]
        --help                       Prints help information
    -V, --version                    Prints version information

ARGS:
    <FILE>    glTF file name
```
Both .gltf and .glb files are supported.
Navigate the scene with the mouse: Rotate with left click + drag, pan with right click + drag, zoom with mouse wheel.

### Example
```
$ curl -O https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Models/master/2.0/Box/glTF-Binary/Box.glb
$ gltf-viewer Box.glb
```

### Headless screenshot generation
Proper headless screenshot generation with the `--headless` flag currently only works on macOS.
To work around that, a Docker setup that uses `xvfb` is provided. Usage examples:
```
# Build docker image and run it with the gltf mounted in a volume.
# The image will be saved next to the gltf file.
./screenshot_docker.sh Box.glb
./screenshot_docker.sh ../models/Box.gltf -w 1920 -h 1080 --count 3 -vv
# Use pre-built docker image from Docker Hub
DOCKER_IMAGE=bwasty/gltf-viewer ./screenshot_docker.sh Box.glb
```

Alternatively, you can also install `xvfb` and use `./run_xvfb.sh` directly (Linux only).


CHANGELOG.md
# Changelog

## [0.4.1] - 2019-01-27
### Fixed
* Ambiguous import compilation error with Rust 1.31 ([#55](https://github.com/bwasty/gltf-viewer/pull/55))
* Crash due to near plane 0 ([#54](https://github.com/bwasty/gltf-viewer/issues/54), [rust-lang/rust#56316](https://github.com/rust-lang/rust/issues/56316))

### Changed
* (internal) upgrade to Rust Edition 2018

## [0.4.0] - 2018-08-22
### Added
* `--straight` CLI parameter: Position camera in front of model if using default camera (i.e. glTF doesn't contain a camera or `--cam-index -1` is passed) ([#49](https://github.com/bwasty/gltf-viewer/pull/49))
* Support more glTF options ([#51](https://github.com/bwasty/gltf-viewer/pull/51)/[#3](https://github.com/bwasty/gltf-viewer/issues/3), now complete apart from animations):
  - `TEXCOORD_1`
  - alpha blending (Alpha Mode, Alpha Cutoff)
    - no depth/transparency sorting yet though, so result isn't always correct
  - all draw modes (Points, Lines, Triangle Strips etc.)

### Fixed
* Rendering of large objects -> dynamic near/far plane ([#31](https://github.com/bwasty/gltf-viewer/pull/31))
* Adding camera bounds cause incorrect bounding box estimation ([#47](https://github.com/bwasty/gltf-viewer/pull/47))
* Minor PBR shader fixes (copied from the reference shader)

## [0.3.1] - 2018-03-16
### Changed
* update `gltf` crate (0.9.3 -> 0.10.1)
  - partially fixes sparse accessor handling (-> https://github.com/gltf-rs/gltf/pull/148, https://github.com/gltf-rs/gltf/issues/154)

### Fixed
* Decode images from data URIs (fully support embedded glTF files, #26)
* Fix non-headless screenshots on high DPI displays (#23)
* Fix panning direction (#25)

## [0.3.0] - 2018-02-23
### Added
* re-add keyboard controls - `WASD` and cursor keys
* remaining glTF camera options:
  - orthographic
  - infinite perspective
* camera CLI parameters:
  - `--cam-index` (choose camera from gltf file)
  - `--cam-pos`
  - `--cam-target`
  - `--cam-fovy` (zoom)
* ambient lighting
* vertex colors (`COLOR_0` vertex attribute)
* `--scene` parameter to choose from multiple scenes in a file

### Changed
* default field of view (fovy) from 60˚ to 75˚

### Fixed
* Initial 'nice' camera position (some models were not properly centered)
* Zooming with real mouse wheel

## [0.2.0] - 2018-02-11
### Changed
* replace `WASD+mouse` navigation with `OrbitControls` based on three.js:
  - rotate: left click + drag
  - pan: right click + drag (still a bit buggy after rotation/zoom)
  - zoom: mouse wheel
* background for screenshots is now transparent

### Added
* `--count` parameter (short `-c`) to generate multiple screenshots, rotating evenly spaced around the object (#10)
* `--headless` parameter for real headless screenshot generation. Unfortunately it only works on macOS. For a workaround using `xvfb`/Docker see [here](https://github.com/bwasty/gltf-viewer#headless-screenshot-generation).
   - Docker images are automatically built for each branch & tag: https://hub.docker.com/r/bwasty/gltf-viewer/


## [0.1.0] - 2017-09-23
* **add PBR shading**
* add screenshot generation via parameter (`-s, --screenshot <FILE>`)
  - Note: currently creates a hidden window, as the "headless" mode of `glutin` did not work as expected
* Determine 'nice' initial camera position from bounding box of scene
  - doesn't work in all cases yet
  - glTF cameras are still ignored
* update [gltf](https://github.com/gltf-rs/gltf) crate to 0.9.2
* **remove URL parameter** (HTTP handling needs to be re-implemented differently after large changes to the `gltf` crate)

## [0.0.3] - 2018-07-28
* add binaries for Win/Linux/macOS to releases
* switch from `glfw-rs` to `glutin`
* add performance logging
* internal refactoring & optimizations

## [0.0.2] - 2018-07-15
* Fix shader handling (`cargo install`ed version didn't work as shaders weren't compiled into the binary).

## [0.0.1] - 2018-11-15
* Initial release. Can display most glTF files, but without any lighting or textures.

---
The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).


src/main.rs
// #![allow(dead_code)]
// #![allow(unused_features)]
// #![feature(test)]
#![allow(clippy::cast_lossless, clippy::cyclomatic_complexity)]

use clap::crate_version;
use cgmath;
use cgmath::Deg;

use clap::{Arg, App, AppSettings};

use log::warn;

use simplelog::{TermLogger, LevelFilter, Config as LogConfig};

mod utils;
mod viewer;
use crate::viewer::{GltfViewer, CameraOptions};

mod shader;
mod controls;
mod framebuffer;
mod macros;
mod importdata;
// TODO!: adapt Source...
// mod http_source;
// use http_source::HttpSource;
mod render;
use crate::render::math::*;

pub fn main() {
    let args = App::new("gltf-viewer")
        .version(option_env!("VERSION").unwrap_or(crate_version!()))
        .setting(AppSettings::UnifiedHelpMessage)
        .setting(AppSettings::DeriveDisplayOrder)
        .before_help("glTF 2.0 viewer\n\nNavigate with the mouse (left/right click + drag, mouse wheel) \
                    or WASD/cursor keys.")
        .arg(Arg::with_name("FILE") // TODO!: re-add URL when fixed...
            .required(true)
            .takes_value(true)
            .help("glTF file name"))
        .arg(Arg::with_name("verbose")
            .long("verbose")
            .short("v")
            .multiple(true)
            .help("Enable verbose logging (log level INFO). Can be repeated up to 3 times to increase log level to DEBUG/TRACE)"))
        .arg(Arg::with_name("screenshot")
            .long("screenshot")
            .short("s")
            .value_name("FILE")
            .help("Create screenshot (PNG)"))
        .arg(Arg::with_name("WIDTH")
            .long("width")
            .short("w")
            .default_value("800")
            .help("Width in pixels")
            .validator(|value| value.parse::<u32>().map(|_| ()).map_err(|err| err.to_string())))
        .arg(Arg::with_name("HEIGHT")
            .long("height")
            .short("h")
            .default_value("600")
            .help("Height in pixels")
            .validator(|value| value.parse::<u32>().map(|_| ()).map_err(|err| err.to_string())))
        .arg(Arg::with_name("COUNT")
            .long("count")
            .short("c")
            .default_value("1")
            .help("Saves N screenshots of size WxH, rotating evenly spaced around the object")
            .validator(|value| value.parse::<u32>().map(|_| ()).map_err(|err| err.to_string())))
        .arg(Arg::with_name("headless")
            .long("headless")
            .help("Use real headless rendering for screenshots (default is a hidden window) [EXPERIMENTAL - see README for details]"))
        .arg(Arg::with_name("straight")
            .long("straight")
            .help("Position camera in front of model if using default camera (i.e. glTF doesn't contain a camera or `--cam-index -1` is passed)"))
        .arg(Arg::with_name("scene")
            .long("scene")
            .default_value("0")
            .help("Index of the scene to load")
            .validator(|value| value.parse::<u32>().map(|_| ()).map_err(|err| err.to_string())))
        .arg(Arg::with_name("CAM-INDEX")
            .long("cam-index")
            .takes_value(true)
            .default_value("0")
            .allow_hyphen_values(true)
            .help("Use the glTF camera with the given index (starting at 0). \n\
                Fallback if there is none: determine 'nice' camera position based on the scene's bounding box. \
                Can be forced by passing -1. \n\
                Note: All other camera options are ignored if this one is given.")
            .validator(|value| value.parse::<i32>().map(|_| ()).map_err(|err| err.to_string())))
        .arg(Arg::with_name("CAM-POS")
            .long("cam-pos")
            .takes_value(true)
            .allow_hyphen_values(true)
            .help("Camera (aka eye) position override as comma-separated Vector3. Example: 1.2,3.4,5.6"))
        .arg(Arg::with_name("CAM-TARGET")
            .long("cam-target")
            .takes_value(true)
            .allow_hyphen_values(true)
            .help("Camera target (aka center) override as comma-separated Vector3. Example: 1.2,3.4,5.6"))
        .arg(Arg::with_name("CAM-FOVY")
            .long("cam-fovy")
            .takes_value(true)
            .default_value("75")
            .help("Vertical field of view ('zoom') in degrees.")
            .validator(|value| value.parse::<u32>().map(|_| ()).map_err(|err| err.to_string())))
        .get_matches();
    let source = args.value_of("FILE").unwrap();

    let width: u32 = args.value_of("WIDTH").unwrap().parse().unwrap();
    let height: u32 = args.value_of("HEIGHT").unwrap().parse().unwrap();
    let count: u32 = args.value_of("COUNT").unwrap().parse().unwrap();

    let scene: usize = args.value_of("scene").unwrap().parse().unwrap();

    let camera_options = CameraOptions {
        index: args.value_of("CAM-INDEX").map(|n| n.parse().unwrap()).unwrap(),
        position: args.value_of("CAM-POS").map(|v| parse_vec3(v).unwrap()),
        target: args.value_of("CAM-TARGET").map(|v| parse_vec3(v).unwrap()),
        fovy: args.value_of("CAM-FOVY").map(|n| Deg(n.parse().unwrap())).unwrap(),
        straight: args.is_present("straight"),
    };

    let log_level = match args.occurrences_of("verbose") {
        0 => LevelFilter::Warn,
        1 => LevelFilter::Info,
        2 => LevelFilter::Debug,
        _ => LevelFilter::Trace
    };

    let _ = TermLogger::init(log_level, LogConfig { time: None, target: None, ..LogConfig::default() });

    let mut viewer = GltfViewer::new(source, width, height,
        args.is_present("headless"),
        !args.is_present("screenshot"),
        camera_options,
        scene);

    if args.is_present("screenshot") {
        let filename = args.value_of("screenshot").unwrap();

        if !filename.to_lowercase().ends_with(".png") {
            warn!("filename should end with .png");
        }
        if count > 1 {
            viewer.multiscreenshot(filename, count)
        } else {
            viewer.screenshot(filename)
        }
        return;
    }

    viewer.start_render_loop();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn print_struct_sizes() {
        // run with `cargo test -- --nocapture`
        println!("Sizes in bytes:");
        println!("Scene:     {:>3}", std::mem::size_of::<render::Scene>());
        println!("Node:      {:>3}", std::mem::size_of::<render::Node>());
        println!("Mesh:      {:>3}", std::mem::size_of::<render::Mesh>());
        println!("Primitive: {:>3}", std::mem::size_of::<render::Primitive>());
        println!("Vertex:    {:>3}", std::mem::size_of::<render::Vertex>());
        println!();
        println!("Option<String>: {:>3}", std::mem::size_of::<Option<String>>());
        println!("String:         {:>3}", std::mem::size_of::<String>());
        println!("Vec<f32>:       {:>3}", std::mem::size_of::<Vec<f32>>());
        println!("Vec<Node>:      {:>3}", std::mem::size_of::<Vec<render::Node>>());
    }

//     extern crate test;
//     use self::test::Bencher;
//     #[bench]
//     fn bench_frame_timer(b: &mut Bencher) {
//         let mut timer = FrameTimer::new("Foobar", 60);
//         b.iter(|| {
//             for _ in 0..60 {
//                 timer.start();
//                 timer.end();
//             }
//         })
//     }
}
